src/storage/ob_partition_parallel_merge_ctx.cpp

```diff
@@ -160,6 +162,473 @@ int ObParallelMergeCtx::init_serial_merge()
+        const int64_t tablet_size = merge_ctx.table_schema_->get_tablet_size();
+
+        int64_t base_macro_blk_cnt = first_sstable_->get_macro_block_ids().count();
+        int64_t num_blks_per_range = tablet_size / OB_FILE_SYSTEM.get_macro_block_size();
+        int64_t num_rows_per_range;
+        if (base_macro_blk_cnt > num_blks_per_range) {
+          num_rows_per_range = first_sstable_->get_total_row_count() / base_macro_blk_cnt * num_blks_per_range;
+        }
+        if (num_rows_per_range <= 0) {
+          num_rows_per_range = 100000;
+        }
+        get_tbl_param.tables_handle_ = &(inc_handle);
+        common::ObExtStoreRange range_to_scan;
+        range_to_scan.get_range().set_whole_range();
+        range_to_scan.get_range().set_start_key(start_key);
+        ObMultipleScanMerge mpl_scan_mrg;
+        mpl_scan_mrg.set_iter_del_row(true);
+        // to open multiple scan merge and generate ranges
+        if (OB_FAIL(mpl_scan_mrg.init(tbl_xs_param, tbl_xs_ctx, get_tbl_param))) {
+          STORAGE_LOG(WARN, "Failed to init multiple scan merge.", K(ret));
+        } else if (OB_FAIL(mpl_scan_mrg.open(range_to_scan))) {
+          STORAGE_LOG(WARN, "Failed to open multiple scan merge.", K(ret));
+        } else if (OB_FAIL(get_ranges_by_iteration(ranges, mpl_scan_mrg, allocator_, num_rows_per_range))) {
+          STORAGE_LOG(WARN, "Failed to get ranges.", K(ret));
+        } else if (OB_UNLIKELY(!contiguous_ranges(ranges))) {
+          ret = OB_ERR_UNEXPECTED;
+          STORAGE_LOG(WARN, "Unexpected bad ranges.", K(ret));
+        } else {
+          STORAGE_LOG(INFO, "Succeeded to get ranges by inc data.", K(table_id), K(range_to_scan.get_range()), K(ranges));
+        }
+      }
+    }
+
+    // revert store context
+    if (store_ctx.mem_ctx_ != NULL) {
+      store_ctx.mem_ctx_->trans_end(true, 0);
+      store_ctx.mem_ctx_->trans_clear();
+      memctx_factory->free(store_ctx.mem_ctx_);
+      store_ctx.mem_ctx_ = NULL;
+    }
+  }
+
+  return ret;
+}
+
+// generate ranges according to both <base_ranges> and <inc_ranges>
+static int combine_ranges(
+    const common::ObIArray<common::ObStoreRange>& base_ranges,
+    const common::ObIArray<common::ObStoreRange>& inc_ranges,
+    common::ObIArray<common::ObStoreRange>& result_ranges)
+{
+  int ret = OB_SUCCESS;
+
+  if (OB_UNLIKELY(base_ranges.count() < 2 || inc_ranges.count() < 2)) {
+    ret = OB_INVALID_ARGUMENT;
+    STORAGE_LOG(WARN, "Invalid args.", K(ret), K(base_ranges), K(inc_ranges));
+  } else {
+    int64_t i = 0, j = 0;
+    common::ObStoreRange range;
+    range.set_right_open();
+    range.set_left_closed();
+    range.get_end_key().set_min();
+    while (OB_SUCC(ret) && i < base_ranges.count() && j < inc_ranges.count()) {
+      const common::ObStoreRowkey& ek1 = base_ranges.at(i).get_end_key();
+      const common::ObStoreRowkey& ek2 = inc_ranges.at(j).get_end_key();
+      const common::ObStoreRowkey *ek;
+      if (ek1 < ek2) {
```

Comment:
the inc ranges is start with the endkey of the base sstable
why we compare the rowkeys here